feat: add cluster health endpoint

mustafadagher opened the pull request:
## Description

Adds a public endpoint to report if the orchestration cluster is `UP`.
Defines `UP` as: There's at least one partition with a healthy leader in the cluster

Returns no content, with the following status codes:
- `204` in case of `UP`
- `503` in case of `DOWN`

## Related issues

closes #20921

mustafadagher committed:
feat: add cluster status endpoint

A public endpoint to report if the orchestration cluster is UP.
Defines `UP` as: There's at least one partition with a healthy leader in the cluster

Returns no content, with the following status codes:
- `204` in case of `UP`
- `503` in case of `DOWN`

mustafadagher committed:
feat: add `/v2/status` to the public unprotected endpoint paths

It's required that the endpoint is available publicly without authentication

mustafadagher committed:
test: add comprehensive tests for StatusController

- Add tests for healthy partition leader scenarios (returns 204 No Content)
- Add tests for unhealthy partition leader scenarios (returns 503 Service Unavailable)
- Add edge case tests for empty partition lists
- Add tests for mixed partition health statuses

mustafadagher committed:
docs: add the cluster status endpoint to the rest api specs

megglos reviewed (approved):
My review is focused on the spec only. I would be against introducing this with a dedicated Status Tag but rather reuse the existing Cluster tag. Pre-approving under the assumption you apply the suggestion.

mustafadagher commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -48,6 +48,7 @@ tags:
   - name: Role
   - name: Setup
   - name: Signal
+  - name: Status
```
Fair point 👍 changed it ✅

deepthidevaki reviewed (approved):
:rocket:  Thanks. Just some minor comments. I'm pre-approving. Feel free to re-request review if needed.

:x: Should we have an integration test as well? Start standalone gateway and brokers 
1. Verify that the status becomes UP 
2. Force no leader for partition - (may be shutdown all brokers) -> Verify that the status is down

mustafadagher commented on zeebe/gateway-rest/src/test/java/io/camunda/zeebe/gateway/rest/controller/StatusControllerTest.java:
```diff
@@ -0,0 +1,181 @@
+
+  @MockitoBean BrokerClient brokerClient;
+  @MockitoBean BrokerTopologyManager topologyManager;
+  @Mock BrokerClusterState clusterState;
```
I can do that, I just unsure what would be the value?
It will also be more difficult to `verify` that we only needed to query the health until the first healthy partition and not all partitions to report healthy (however, you can argue that this is implementation details too)

mustafadagher committed:
feat: add cluster status command to java client

mustafadagher committed:
test: add integration tests to verify cluster status endpoint

- reports UP when a cluster is healthy
- reports DOWN when all brokers are stopped
- reports UP when only 1 partition has Healthy leader

deepthidevaki commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusterStatusTest.java:
```diff
@@ -0,0 +1,123 @@
+        final var broker = CLUSTER.brokers().get(MemberId.from(String.valueOf(i)));
+        broker.stop();
+      }
+      final var broker1 = CLUSTER.brokers().get(MemberId.from("0"));
+      final var broker2 = CLUSTER.brokers().get(MemberId.from("1"));
+      broker1.stop();
+      broker2.stop();
```
:x: This is redundant with the above loop.

mustafadagher replied:
oh sorry, I meant to remove the loop and replace it with this for clarity, and forgot both :D

deepthidevaki commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusterStatusTest.java:
```diff
@@ -0,0 +1,123 @@
+      final var initialResponse = client.newStatusRequest().send().join();
+      assertThat(initialResponse.isHealthy()).isTrue();
+
+      // when - stopping 2 out of 3 brokers, leaving at least one partition with a leader
```
:x: Since replication factor is 3, if you stop 2 brokers then there won't be a leader for any partition. The test is likely passing now because the update that leader does not exist is not received yet by the gateway.

For this test to work, I would suggest to set the replicationFactor to 1. For the other two tests it does not matter what the replicationFactor is. 

To ensure we are checking the status correctly, may be wait until two partitions has no leader before checking the status is still healthy.

deepthidevaki commented on zeebe/qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/ClusterStatusTest.java:
```diff
@@ -0,0 +1,123 @@
+ * <p>Tests that the cluster status correctly reports healthy when at least one partition has a
+ * healthy leader, and unhealthy when no partitions have healthy leaders.
+ */
+@Testcontainers
```
:x: This test is not using test containers, right?

deepthidevaki commented on zeebe/gateway-protocol/src/main/proto/rest-api.yaml:
```diff
@@ -71,6 +71,20 @@ paths:
+      description: Checks the health status of the cluster by verifying if there's at least one partition with a healthy leader.
+      responses:
+        "204":
+          description: The cluster is healthy and has at least one partition with a healthy leader.
```
:wrench: May be replace "healthy" with UP?  - `The cluster is UP and has at least one partitions with a healthy leader.`

mustafadagher replied:
💡 On that note, I'll also change the java-client response to return an enum (UP/DOWN) instead of a boolean `isHealthy`.

deepthidevaki reviewed (approved):
:rocket: Thanks for writing the integration test.

:x: Please fix the test before merging.

backport-action commented:
Successfully created backport PR for `release-8.8.0-alpha8`:
- #37327

mustafadagher commented:
/backport
